### DISCUSSION TITLE
API query returns list, repeater needs a refresh after each local modification (workaround)

### DISCUSSION THREAD

**Discussion by bjornclauw** [Budibase/budibase#6766] [Ideas]
Hi,

I use a REST query to populate a repeater with 'tasks'.

I have several actions I can perform on each task. 
For each of these tasks I have a button which sends a PUT request to edit the item on the remote server.

**The problem:** 
Every time I a make a local modification, I have to refresh my data provider to see my last modification.
(which results in a lot of API calls and a lot small delays for the user)

**Workaround:**
1. In my repeater block I add a new container
1. Inside this container, I add a form
2.  In the form I add a textfield named: `hidden_value_deleted`
3. In the custom CSS of this textfield I add: `display: none;`
4. Button action: After my PUT query I add: `Update field value` and set value for `hidden_value_deleted` to `1`.
5. Now i add a condition to my container: `if 'hidden_value_deleted' IS NOT empty  > HIDE container`

-> So when the PUT query results in succes my changes are shown 'offline' so I do not need to refresh after each edit.
(Also useful when changing a value, ...)

**Idea:**
Would it be possible to implement this into BB? 
Maybe by adding an ACTION to the button component that allows us to edit/modify a bindable value (inside the repeater).

(Let's say we pull a list of items with a `name` and a `status` attribute. We have a button that makes a PUT request to change `status` from TODO to DONE. When API call is succesfull `{{repeater.status}}` is set to the value DONE. 
(this would be our new action in 'On-click button' menu).
> Because we have set a filter in our data provider to filter on TODO, the current task will dissapear from our list. (without the need to refresh), or we can use a condition and change the color of a component, show or hide text, update text to new value, ... in relation to the bindable value.

Hope this makes some sense, I included a demo app with working API request.
(note, the PUT request on the button is omitted because I use a public API that I cannot write to in this example.)

![image](https://user-images.githubusercontent.com/32811735/179404225-0779b2c2-375d-4c5b-ac08-ab75197560fa.png)
![image](https://user-images.githubusercontent.com/32811735/179404992-a2deacc7-8aea-4dfc-9e3b-531316746fc2.png)

[api_test-export-1658067284236.txt](https://github.com/Budibase/budibase/files/9127874/api_test-export-1658067284236.txt)

**Comment by melohagan** [comment]
Hi @bjornclauw 

Thank you for the detailed idea! Something that I've wanted to do in the past which I think is applicable to this scenario is to use <a href="https://docs.budibase.com/docs/app-state">App State</a> as a data source for a Data Provider.

If you updated the app state, it would be much more efficient to refresh the data from in-memory than making an external API call.

**Reply by bjornclauw** [reply]
Hi @melohagan,

that sounds interesting, would it look something like this?
![image](https://user-images.githubusercontent.com/32811735/179497640-b2f12574-ff5f-41b0-928b-a30087335a46.png)

So our data source remains (=api call or database or ...) and we can add as many 'App States' as we want from directly inside the data provider.

>  So: each item in the data source gains a variable, per 'App State' we add in the data provider.

-> if this is what you mean, this would be a really powerfull addition codewise. Not only for optimizing API calls, but also for automating UI changes/updates. 

(especially returning feedback from a button after a button-press to the corresponding 'item' in a repeater list. Only possible for me with the workaround mentioned above -_without reloading the datasource after each edit_)

**Reply by melohagan** [reply]
I was thinking more of a nested scenario:

> Data Provider (REST Query datasource)
> > Data Provider (App State datasource)
> > > Repeater Block

Then your button would update the App State, which would only cause the App State Data Provider to be refreshed, and would not re-trigger the REST API call.
